### PR TITLE
Add logic to check for incompatible parameters

### PR DESCRIPTION
**Pull request by rvhonorato**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [x] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [x] Your code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

This PR adds a logic to check for parameter incompatibilities. 

For example; with #918 I added a new parameter called `less_io` however you cannot use this parameter together with `mode=batch`.

So now you can define a new `incompatible` property of a given parameter in the `haddock.modules.defaults.yaml`;

```yaml
less_io:
  # ...
  incompatible:
    mode: batch
```

Which will then be validated when the run is being setup.

### CODE REVIEW

File: src/haddock/modules/defaults.yaml [hunk 1]
```diff
@@ -148,3 +148,5 @@ less_io:
     a network file system where I/O operations are slow.
   group: "execution"
   explevel: easy
+  incompatible:
```
Comment (amjjbonvin) [inline src/haddock/modules/defaults.yaml:151]: Nice!  Will the syntax allow for multiple entries under the incompatible category? I assume yes

**Review by amjjbonvin — APPROVED** [review]

File: src/haddock/gear/yaml2cfg.py [hunk 2]
```diff
@@ -214,3 +215,28 @@ def flat_yaml_cfg(cfg: ParamMap) -> ParamDict:
         new[param] = new_value
 
     return new
+
+
+def find_incompatible_parameters(yaml_file: Path) -> dict[str, dict[str, str]]:
```
Comment (VGPReys) [inline src/haddock/gear/yaml2cfg.py:220]: I would change the return type to: dict[str, ParamDict]

File: src/haddock/gear/prepare_run.py [hunk 3]
```diff
@@ -76,6 +77,7 @@
     modules_category,
     modules_names,
     non_mandatory_general_parameters_defaults,
+    incompatible_params,
```
Comment (VGPReys) [inline src/haddock/gear/prepare_run.py:80]: I would change the name of that guy to: incompatible_defaults_params

File: src/haddock/modules/__init__.py [hunk 4]
```diff
@@ -52,25 +53,28 @@
...
 # module where the module definition overwrites global definition. Not all
 # modules will use these parameters. It is the responsibility of the module to
 # extract the parameters it needs.
 # the config file is in modules/defaults.cfg
-non_mandatory_general_parameters_defaults = read_from_yaml_config(modules_defaults_path)  # noqa : E501
+non_mandatory_general_parameters_defaults = read_from_yaml_config(
+    modules_defaults_path
+)  # noqa : E501
+
+incompatible_params = find_incompatible_parameters(modules_defaults_path)
```
Comment (VGPReys) [inline src/haddock/modules/__init__.py:68]: BIS: I would change the name of that guy to: incompatible_defaults_params

File: src/haddock/gear/prepare_run.py [hunk 5]
```diff
@@ -940,6 +944,27 @@ def validate_module_names_are_not_misspelled(params: ParamMap) -> None:
     return
 
 
+def validate_parameters_are_not_incompatible(params: ParamMap) -> None:
+    """Validate parameters are not incompatible.
+
+    Args:
```
Comment (VGPReys) [inline src/haddock/gear/prepare_run.py:950]: In the rest of the haddock3 code, we are using the `numpy` docstring format

**Review by VGPReys — DISMISSED** [review]

**Comment by sverhoeven** [comment]
In clustrmsd module the value of criterion parameter determines which other module parameters are ignored see https://github.com/haddocking/haddock3/blob/a11a7c37a96b12d203717941607daf36bf966eaa/src/haddock/modules/analysis/clustrmsd/defaults.yaml#L50 . Could we use incompatible parameters there as well? Or use clustrrmsd approach here?

**Comment by VGPReys** [comment]
> In clustrmsd module the value of criterion parameter determines which other module parameters are ignored see
> 
> https://github.com/haddocking/haddock3/blob/a11a7c37a96b12d203717941607daf36bf966eaa/src/haddock/modules/analysis/clustrmsd/defaults.yaml#L50
> . Could we use incompatible parameters there as well? Or use clustrrmsd approach here?

I don't know, because they have default values, and only in some case (based on `criterion` value) some parameters will be used but not the other ones.

File: src/haddock/modules/defaults.yaml (hunk 1 above)
Comment (rvhonorato) [reply src/haddock/modules/defaults.yaml:151]: Default `yaml` format syntax applies here

**Review by amjjbonvin — DISMISSED** [review]
We will have to remember to edit the default.yaml files when needed to define incompatible parameters, but now at least the machinery to check those is in place - nice!

**Review by VGPReys — APPROVED** [review]

**Review by amjjbonvin — APPROVED** [review]

**Review by AnnaKravchenko — APPROVED** [review]

**Pull request merged**